File: week3/phone_book/phone_book.cpp
#include <iostream>
#include <vector>
#include <string>

struct Query{
    std::string procedure, name;
    int number; 
};

std::vector<Query> readQueries(){
    int n;
    std::cin >> n;
    std::vector<Query> queries(n);
    for (int i = 0; i < n; i++){
        std::cin >> queries[i].procedure;
        if (queries[i].procedure == "add"){
            std::cin >> queries[i].number;
            std::cin >> queries[i].name;
        }
        else{
            std::cin >> queries[i].number;
        }
    }
    return queries;
}

std::vector<std::string> processQueries(std::vector<Query> queries){
    std::vector<std::string> response;
    std::vector<Query> contacts(10000000);
    for (int i = 0; i < queries.size(); i++){
        if (queries[i].procedure == "add"){
            contacts[queries[i].number] = queries[i];
        }
        else if (queries[i].procedure == "find"){
            if (contacts[queries[i].number].name != ""){
                response.push_back(contacts[queries[i].number].name);
            }
            else{
                response.push_back("not found");
            }
        }
        else if (queries[i].procedure == "del"){
            contacts[queries[i].number].number = 0;
            contacts[queries[i].number].name = "";
        }
        else{
            std::cout << "wrong procedure, please only choose find, add or del" << std::endl;
        }
    }
    return response;
}


void readResponse(std::vector<std::string> response){
    for (auto const& res: response){
        std::cout << res << std::endl;
    }
}

int main(){
    readResponse(processQueries(readQueries()));
    return 0;
}

File: week3/phone_book/phone_book.py
# python3
class Query:
    def __init__(self, query):
        self.procedure = query[0]
        self.number = int(query[1])
        if self.procedure == "add":
            self.name = query[2]

def ReadQueries():
    n = int(input())
    return [Query(input().split()) for i in range(n)]

def ProcessQueries(queries):
    response = []
    contacts = [None for i in range(0, 10000000)]
    for query in queries:
        if query.procedure == "add":
            contacts[query.number] = query
        elif query.procedure == "find":
            if contacts[query.number] != None:
                response.append(contacts[query.number].name)
            else:
                response.append("not found")
        elif query.procedure == "del":
            contacts[query.number] = None
        else:
            print("wrong procedure, please only chosse find, add or del")
    return response

def ReadResponse(response):
    for res in response:
        print(res)

if __name__ == '__main__':
    ReadResponse(ProcessQueries(ReadQueries()))
